opentelemetry-api/src/opentelemetry/trace/__init__.py

```diff
@@ -116,6 +119,35 @@
 logger = getLogger(__name__)
 
 
+class _AgnosticContextManager(
+    contextlib._GeneratorContextManager
+):  # pylint: disable=protected-access
```

Comment:
I'm not sure if it's a good to rely on this internal class and `self._recreate_cm()` explicitly. We might be better off sticking to the public API. I think you could probably do it with composition instead of inheritance to avoid this